metal_header/sifive_ccache0.c++

```diff
@@ -13,76 +13,42 @@ void sifive_ccache0::include_headers() {
             [&](node n) { emit_include(compat_string); });
 }
 
-void sifive_ccache0::declare_inlines() {
-  Inline *func;
-  std::list<Inline *> extern_inlines;
-  int count = 0;
+void sifive_ccache0::create_defines() {
 
   dtb.match(std::regex(compat_string), [&](node n) {
-    if (count == 0) {
-      func = create_inline_dec("control_base", "uintptr_t",
-                               "struct metal_cache *cache");
-      extern_inlines.push_back(func);
+    uint32_t i0 = 0, i1 = 0, i2 = 0, i3 = 0;
+    if (n.field_exists("interrupts")) {
+      /* Get interrupt source IDs */
+      i0 = n.get_fields<uint32_t>("interrupts")[0];
+      i1 = n.get_fields<uint32_t>("interrupts")[1];
+      i2 = n.get_fields<uint32_t>("interrupts")[2];
+      i3 = n.get_fields<uint32_t>("interrupts")[3];
     }
-    count++;
-  });
-  os << "\n";
-  os << "/* --------------------- sifive_ccache0 ------------ */\n";
-  while (!extern_inlines.empty()) {
-    func = extern_inlines.front();
-    extern_inlines.pop_front();
-    emit_inline_dec(func, "sifive_ccache0");
-    delete func;
-  }
-  os << "\n";
-}
 
-void sifive_ccache0::define_inlines() {
-  Inline *func;
-  std::list<Inline *> extern_inlines;
-
-  int count = 0;
-  dtb.match(std::regex(compat_string), [&](node n) {
-    if (count == 0) {
-      func = create_inline_def("control_base", "uintptr_t",
-                               "(uintptr_t)cache == (uintptr_t)&__metal_dt_" +
-                                   n.handle(),
-                               platform_define(n, METAL_BASE_ADDRESS_LABEL),
-                               "struct metal_cache *cache");
-      add_inline_body(func, "else", "0");
-      extern_inlines.push_back(func);
-    }
-    count++;
-  });
-  os << "\n";
-  os << "/* --------------------- sifive_ccache0 ------------ */\n";
-  while (!extern_inlines.empty()) {
-    func = extern_inlines.front();
-    extern_inlines.pop_front();
-    emit_inline_def(func, "sifive_ccache0");
-    delete func;
-  }
-  os << "\n";
-}
+    emit_def("METAL_SIFIVE_CCACHE0_INTERRUPTID0", std::to_string(i0));
+    emit_def("METAL_SIFIVE_CCACHE0_INTERRUPTID1", std::to_string(i1));
+    emit_def("METAL_SIFIVE_CCACHE0_INTERRUPTID2", std::to_string(i2));
+    emit_def("METAL_SIFIVE_CCACHE0_INTERRUPTID3", std::to_string(i3));
```

Comment:
The number of interrupt are not necessary 4 all the time, and is based on the number of l2 banks.
You will need to count the number of interrupts and create the appropriate number of IDs